File: solutions/python/say/1/say.py
def say(number):
    if number < 0 or number > 999_999_999_999:
        raise ValueError("input out of range")
    basics = ["zero", "one", "two", "three", "four", "five", "six",
             "seven", "eight", "nine", "ten", "eleven", "twelve",
             "thirteen", "fourteen", "fifteen", "sixteen", "seventeen",
             "eighteen", "nineteen"]
    tens = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
    thousands = ["thousand", "million", "billion"]

    if number < 20:
        return basics[number]
    elif number < 100:
        return f"{tens[number//10 - 2]}{"" if number % 10 == 0 else f"-{basics[number % 10]}" }"
    elif number < 1000:
        return f"{basics[number // 100]} hundred{"" if number % 100 == 0 else f" {say(number % 100)}"}"
    else:
        for i in range(len(thousands)-1, -1,-1):
            divider = 10 ** ((i+1)*3)
            if number >= divider:
                return f"{say(number // divider)} {thousands[i]}{"" if number % divider == 0 else f" {say(number % divider)}"}"
        return ""

File: solutions/cpp/say/1/say.cpp
#include "say.h"

#include <stdexcept>
#include <vector>

static const std::vector<std::string> basics = {
    "zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
    "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
};

static const std::vector<std::string> tens = {
    "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
};

static const std::vector<std::string> thousands = {"thousand", "million", "billion"};

namespace say {
    std::string in_english(const int64_t number) {
        if (number < 0 || number > 999999999999) {
            throw std::domain_error("Error");
        }
        if (number < 20) {
            return basics[number];
        }
        if (number < 100) {
            return tens[number /10 - 2] + (number % 10 == 0 ? "": "-" + basics[number % 10]);
        }
        if (number < 1000) {
            return basics[number /100] + " hundred" + (number % 100 == 0 ? "": " " + in_english(number % 100));
        }
        for (int i = 2; i >= 0; i--) {
            int64_t divider = 1;
            for (int j = 0; j < (i + 1) * 3; j++) {
                divider *= 10;
            }
            if (number >= divider) {
                return in_english(number/divider) + " " + thousands[i] + (number % divider == 0 ? "": " " + in_english(number%divider));
            }
        }
        return "";
    }

}  // namespace say
